proposals/002-mdm-profiles-sync.md

```diff
@@ -0,0 +1,58 @@
+    Built-in profiles that are included with fleet should use a hard coded identifier so that we can update these profiles.
+    This may be required when new options are added or bug fixes are required to an existing profile
+    - `payload_uuid`: this should change whenever an existing profile is updated
+        - note, this doesn't actually need to be a uuid
+    - `version`: isn't part of the actually configuration profile. Used to determine what the latest version of a profile is. For simplicity, this can be an integer that is incremented from 1.
+        - note, this is *not* the payload_version which is used by apple to track the format version of the profile, usually `1`.
+    - `raw`: The complete raw configuration profile as xml (mobileconfig)
+    - example query to get the latest version of all the profiles
+        ```
+        SELECT p1.*
+        FROM
+            mdm_apple_configuration_profiles p1
+            LEFT JOIN mdm_apple_configuration_profiles p2
+                  ON p1.payload_identifier = p2.payload_identifier AND p1.version < p2.version
+        WHERE
+            p2.version is NULL
+        ```
+    - How do we handle deleted/disabled profiles? What about builtin profiles?
+        - soft delete ie deleted_at. Need to clean these up periodically?
+        - delete the profile, and allow `host_mdm_apple_configuration_profiles.payload_identifier` to be orphaned
+
+- Store the current profiles installed on mdm enrolled hosts in a `host_mdm_apple_configuration_profiles` table
+    - `host_id`, `payload_identifier`, `payload_uuid`
+    - Populate whenever a `InstallProfile` or periodic `ProfileList` MDM command is run on a host
+
+## Sync
+
+- backend terminology
+    - current state: The profiles and their specific versions that are installed on devices
+    - desired state: The latest profiles and their specific versions
+    - drift: Differnce between the current state and the desired state
+    - refresh: detect drift and reconcile by installing/removing profiles
+- frontent terminology
+    - applied: host has all profiles with the latest version installed.
+    - pending: drift has been detected on the host eg a new profile was created in fleet which needs to be installed on the host
+    - failed: there was an error with the profile itself, or installing the profile on the host
+
+- To avoid issues when syncing profiles eg side effects, we should only install/remove profiles that have actually changed
+    - detect drift/out
+- Add a mechanism to refresh the list of profiles installed on an individual host manually and sync the desired profiles, similar to the osquery refetch host mechanism
+    - helps in troubleshooting
+    - able to quickly test changes ie during development
+- To avoid unexpected behaviour and to increase robustness, periodically issue a `ProfileList` MDM command on hosts
```

Comment:
I have been thinking about this, can we do a `ProfileList` each time we need to know which profiles are installed? it's more network overhead but can save us a lot of work and problems.

Reply:
When you say "each time", do you mean before we install/remove profiles on the devices? The problem with doing a `ProfileList` immediately before install/remove profiles is that it would mean devices need to be online whenever our scheduled job runs that syncs profiles.

In a perfect world, we wouldn't need to do a `ProfileList` command at all because we would know the current state of a host by looking at all the `InstallProfile` commands that have been successfully run.

Reply:
> When you say "each time", do you mean before we install/remove profiles on the devices? 

yes, sorry for not being clear on that

> The problem with doing a ProfileList immediately before install/remove profiles is that it would mean devices need to be online whenever our scheduled job runs that syncs profiles.

but, this is true regardless, we need the device to be online in order to install/delete profiles. It would mean an extra roundtrip, and the device could go offline between the `ProfileList` and the `Install/DeleteProfile` command, but maybe it's OK?

Reply:
we discussed about this offline, and the conclusions are:

1. Due to the async nature of MDM, we'd have to have a job running at very short intervals in order to issue an `InstallProfile` after the `ProfileList`, it's better to trust the state we have in the database, and sync from time to time, just in case
2. This syncing can be tackled separately later